ledger/participant-integration-api/src/main/scala/platform/indexer/ha/PreemptableSequence.scala

```diff
@@ -0,0 +1,212 @@
+  */
+trait SequenceHelper {
+
+  /** Register at any point in time a synchronous release function,
+    * which will be ensured to run before the completion future of the handle completes.
+    *
+    * @param release the release lambda
+    */
+  def registerRelease(release: => Unit): Unit
+
+  /** Wrap a CBN (lazy) Future, so it is only started if the PreemptableSequence is not yet aborted/shut down.
+    *
+    * @param f The lazy Future
+    * @return the wrapped future
+    */
+  def goF[T](f: => Future[T]): Future[T]
+
+  /** Wrap a CBN (lazy) synchronous function in a Future, which is only started if the PreemptableSequence is not yet aborted/shut down.
+    *
+    * @param body The lazy synchronous body
+    * @return the wrapped future
+    */
+  def go[T](body: => T): Future[T]
+
+  /** Wrap a synchronous call into a Future sequence, which
+    * - will be preemptable
+    * - will retry to execute the body if Exception-s thrown
+    *
+    * @return the preemptable, retrying Future sequence
+    */
+  def retry[T](waitMillisBetweenRetries: Long, maxAmountOfRetries: Long = -1)(
+      body: => T
+  ): Future[T]
+
+  /** Delegate the preemptable-future sequence to another Handle
+    * - the completion Future future of the PreemptableSequence will only finish after this Hanlde finishes,
+    *   and previously registered release functions all completed
+    * - KillSwitch events will be replayed to this handle
+    * - In case of abort/shutdown the PreemptableSequence's completion result will conform to the KillSwitch usage,
+    *   not to the completion of this handle (although it will wait for it naturally)
+    *
+    * @param handle The handle to delegate to
+    * @return the completion of the Handle
+    */
+  def merge(handle: Handle): Future[Unit]
+
+  /** The handle of the PreemptableSequence. This handle is available for sequence construction as well.
+    * @return the Handle
+    */
+  def handle: Handle
+}
+
+// these family of KillSwitch-es enable the behavior of recording the usage of the KillSwitch
+// - Shutdown always wins: in scenarios like multiple abort and then a shutdown will always capture a shutdown,
+//   even if additional aborts arrive after the shutdown. This is needed so that graceful-shutdown can stop possible
+//   recovery scenarios.
+// - Always the last abort wins.
+trait UsedKillSwitch extends KillSwitch {
+  override def shutdown(): Unit = ()
+  override def abort(ex: Throwable): Unit = ()
+}
+case object ShutDownKillSwitch extends UsedKillSwitch
+case class AbortedKillSwitch(ex: Throwable, _myReference: AtomicReference[KillSwitch])
+    extends CaptureKillSwitch(_myReference)
+class CaptureKillSwitch(myReference: AtomicReference[KillSwitch]) extends KillSwitch {
+  override def shutdown(): Unit = myReference.set(ShutDownKillSwitch)
+  override def abort(ex: Throwable): Unit = myReference.set(AbortedKillSwitch(ex, myReference))
+}
```

Comment:
@nmarton-da : what is this code still used for?

Reply:
:man_facepalming: removed